src/discord_bot/bot.cpp

```diff
@@ -28,22 +181,170 @@ bot::bot(const bot_context &context)
 }
 
-bot::~bot() {}
+bot::~bot() {
+	std::scoped_lock lock(this->gateway_mutex);
+	if (!this->started) {
+		return;
+	}
+
+	if (this->stopping) {
+		return;
+	}
+
+	if (this->gateway_active) {
+		this->ws_client->close();
+	}
+}
+
+void bot::run() {
+	{
+		std::scoped_lock lock(this->gateway_mutex);
+		if (this->started) {
+			throw std::runtime_error("bot is already running");
+		}
+	}
+
+	this->logger->log(log::level::INFO, "Starting up bot.");
+
+	bool retry;
+	do {
+		std::unique_lock lock(this->gateway_mutex);
+		if (this->stopping) {
+			break;
+		}
+
+		this->logger->log(log::level::DEBUG, "Setting up zlib stream inflator.");
+		zlib_inflate_context z_context;
+		this->z_inflator = std::make_shared<zlib_inflate>(z_context);
+
+		this->logger->log(log::level::DEBUG, "Setting up websocket client.");
+		websocket_endpoint_context ws_context;
+		ws_context.logger       = this->logger;
+		ws_context.tls_ver      = this->tls_ver;
+		ws_context.ca_directory = this->ca_directory;
+		ws_context.http_proxy   = this->http_proxy;
+		this->ws_client         = std::make_shared<websocket_client>(ws_context);
+
+		this->ws_client->set_open_handler([this, &lock]() {
+			this->started        = true;
+			this->gateway_active = true;
+			lock.unlock();
+		});
+
+		this->ws_client->set_fail_handler([this, &lock]() {
+			this->started = true;
+			lock.unlock();
+		});
+
+		this->ws_client->set_close_handler([this, &lock]() {
+			lock.lock();
+			this->gateway_active = false;
+			lock.unlock();
+		});
+
+		this->ws_client->set_message_handler(std::bind(&bot::handle_message, this, std::placeholders::_1));
+
+		// This will block until either the client or server terminates the connection.
+		this->ws_client->connect_and_run(this->ws_url);
+
+		{
+			// Stop heartbeating.
+			std::scoped_lock lock(this->heartbeat_mutex);
+			this->stop_heartbeating.notify_one();
+		}
+
+		websocket_status status;
+		try {
+			status = this->ws_client->get_close_status();
+			this->logger->log(log::level::DEBUG,
+			    fmt::format("Websocket connection closed with code={}; message={}", status.code, status.message));
+		} catch (std::exception &e) {
+			this->logger->log(log::level::ERROR, fmt::format("Error closing connection: {}.", e.what()));
+		}
+
+		// Determine whether to retry or not.
+		switch (status.code) {
+		case close_event_code::unknown_opcode:
+		case close_event_code::decode_error:
+		case close_event_code::not_authenticated:
+		case close_event_code::already_authenticated:
+		case close_event_code::invalid_api_version:
+		case close_event_code::invalid_intents:
+		case close_event_code::disallowed_intents: {
+			// This may be a problem with the library, so don't retry to be safe.
+			this->logger->log(log::level::ERROR, "Logic error; Eelbot Framework library version may be out of date.");
+			retry = false;
+			break;
+		}
+		case close_event_code::authentication_failed:
+		case close_event_code::invalid_shard:
+		case close_event_code::sharding_required: {
```

Comment:
Why aren't cases like this (and the one below logged), you log the first set of unrecoverable errors

Reply:
Good question. I think it was because the connection closing is already logged before. The above cases can provide additional info about what the issue might be so it logs an additional line but that can't be said for these cases so there's no further info to provide.